source/stun_deserializer.c

```diff
@@ -0,0 +1,868 @@
+        }
+    }
+
+    if( result == STUN_RESULT_OK )
+    {
+        if( pAttribute->attributeType == STUN_ATTRIBUTE_TYPE_FINGERPRINT )
+        {
+            pCtx->attributeFlag |= STUN_FLAG_FINGERPRINT_ATTRIBUTE;
+        }
+        if( pAttribute->attributeType == STUN_ATTRIBUTE_TYPE_MESSAGE_INTEGRITY )
+        {
+            pCtx->attributeFlag |= STUN_FLAG_INTEGRITY_ATTRIBUTE;
+        }
+
+        /* Read attribute length. */
+        STUN_READ_UINT16( &( pAttribute->attributeValueLength ),
+                          ( uint8_t * ) &( pCtx->pStart[ pCtx->currentIndex + STUN_ATTRIBUTE_HEADER_LENGTH_OFFSET ] ) );
+
+        /* Check that we have enough data to read attribute value. */
+        if( STUN_REMAINING_LENGTH( pCtx ) < STUN_ATTRIBUTE_TOTAL_LENGTH( pAttribute->attributeValueLength ) )
+        {
+            result = STUN_RESULT_MALFORMED_MESSAGE;
+        }
+    }
+
+    if( result == STUN_RESULT_OK )
+    {
+        pAttribute->pAttributeValue = &( pCtx->pStart[ pCtx->currentIndex + STUN_ATTRIBUTE_HEADER_VALUE_OFFSET ] );
+
+        pCtx->currentIndex += STUN_ATTRIBUTE_TOTAL_LENGTH( STUN_ALIGN_SIZE_TO_WORD( pAttribute->attributeValueLength ) );
+    }
+
+    return result;
+}
+/*-----------------------------------------------------------*/
+
+StunResult_t StunDeserializer_ParseAttributeErrorCode( const StunAttribute_t * pAttribute,
+                                                       uint16_t * errorCode,
+                                                       char ** errorPhrase )
+{
+    StunResult_t result = STUN_RESULT_OK;
+    uint8_t class, errorNumber;
+    uint16_t errorPhaseLength = pAttribute->attributeValueLength - STUN_ERROR_CODE_PACKET_ERROR_PHRASE_OFFSET;
+
+    if( ( pAttribute == NULL ) ||
+        ( errorCode == NULL ) ||
+        ( pAttribute->pAttributeValue == NULL ) ||
+        ( errorPhaseLength <= 0 ) ||
+        ( pAttribute->attributeType != STUN_ATTRIBUTE_TYPE_ERROR_CODE ) )
+    {
+        result = STUN_RESULT_BAD_PARAM;
+    }
+
+    if( result == STUN_RESULT_OK )
+    {
+        class = pAttribute->pAttributeValue[STUN_ERROR_CODE_PACKET_ERROR_CLASS_OFFSET];
+        errorNumber = pAttribute->pAttributeValue[STUN_ERROR_CODE_PACKET_ERROR_CODE_OFFSET];
+
+        *errorCode = STUN_GET_ERROR( class,
+                                     errorNumber );
+        *errorPhrase = ( uint8_t * ) &( pAttribute->pAttributeValue[STUN_ERROR_CODE_PACKET_ERROR_PHRASE_OFFSET] );
+    }
+    return result;
+}
+
+/*-----------------------------------------------------------*/
+
+StunResult_t StunDeserializer_ParseAttributeChannelNumber( const StunContext_t * pCtx,
+                                                           const StunAttribute_t * pAttribute,
+                                                           uint16_t * channelNumber,
+                                                           StunAttributeType_t attributeType )
+{
+    StunResult_t result = STUN_RESULT_OK;
+
```

Comment:
pCtx is not used in this function, should it be passed as an argument?

Reply:
same as above